### PR TITLE
add calculate function

### PR DESCRIPTION
**Pull request by yuweiii**

### CODE REVIEW

File: pyfer/functions.py [hunk 1]
```diff
@@ -61,14 +61,26 @@ def calculate(data, stat="mean"):
...
     pd.DataFrame:
         Summarized data. Each row contains the summary statistic for a given resample.
     '''
-    return
+    if not isinstance(data, pd.DataFrame):
+        raise TypeError("Input should be a Pandas DataFrame")
+    if stat=="Mean":
+        raise TypeError("Input is incorrect. Did you mean 'mean'?")
+    if stat=="Median":
+        raise TypeError("Input is incorrect. Did you mean 'median'?")
```
Comment (GabrielBogo) [inline pyfer/functions.py:88]: This checks are nice (tidyverse uses them all the time!) but they are not sufficient. What if someone types "meedian"? One test for this function is still failing (see "test_statistic_not_implemented") and should be addressed here, additionally.

File: pyfer/functions.py [hunk 2]
```diff
@@ -61,14 +61,26 @@ def calculate(data, stat="mean"):
...
+    if stat=="mean":
+        data['stat']=""
+        for i in range(1, data['sample_id'].max()+1):
+            stat_per_sample=data['response'][data['sample_id']==i].mean()
+            data.loc[data['sample_id']==i,'stat'] = stat_per_sample
+
```
Comment (GabrielBogo) [inline pyfer/functions.py:94]: I understand the logic of this, but the shape of the output should be reduced by `calculate`. See the test `test_shape_output`, it's still failing. Try the pandas `groupby` function: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html

File: pyfer/functions.py [hunk 3]
```diff
@@ -61,14 +61,26 @@ def calculate(data, stat="mean"):
...
     '''
-    return
+    if not isinstance(data, pd.DataFrame):
+        raise TypeError("Input should be a Pandas DataFrame")
+    if stat=="Mean":
+        raise TypeError("Input is incorrect. Did you mean 'mean'?")
+    if stat=="Median":
+        raise TypeError("Input is incorrect. Did you mean 'median'?")
+    if stat=="mean":
+        data['stat']=""
```
Comment (GabrielBogo) [inline pyfer/functions.py]: You called the column 'stat', but in the specification it was written `mean`. Another test or this function is failing, see "test_output_statistic".

I think calling it `stat` is reasonable, actually. But then we'll have to change the tests.

**Review by GabrielBogo — COMMENTED** [review]
-

File: pyfer/functions.py (hunk 1 above)
Comment (yuweiii) [reply pyfer/functions.py:88]: Hi Gabriel. About  the "meedian", I was thinking that, compared with "Median", "meedian" is a mistake that can be more easily detected. That was why I chose "Median." About the "test_statistic_not_implemented", I am taking look at the "test_statistic_not_implemented" in the test right now. will implement it later.

File: pyfer/functions.py (hunk 2 above)
Comment (yuweiii) [reply pyfer/functions.py:94]: Hi Gabriel, I misunderstood the shape of the output as the same number of rows as generate and one more column that generate. Will implement this

File: pyfer/functions.py (hunk 3 above)
Comment (yuweiii) [reply pyfer/functions.py]: I am confused about this. In the documentation of calculate, it says one of the parameters is ` stat: string. "mean" (default) or "median".` I thought it means the name of parameter is stat, and mean is the input of `stat`?

File: pyfer/functions.py (hunk 1 above)
Comment (makka3) [reply pyfer/functions.py:88]: Why not try `if not (stat == "mean"): raise TypeError("This function only takes "mean" as input")` or something to that effect. This way you can cover all the cases that are not `mean` in one go.

**Comment by yuweiii** [comment]
Just updated this branch.

**Comment by yuweiii** [comment]
Okay. Since @GabrielBogo has reviewed it and @GabrielBogo also made some changes himself, I am going to merge his changes as long as mine.

**Pull request merged**